File: ticket-booking-admin/src/services/promoCodeService.ts
import api from './api';
import { PromoCode, PromoCodeRequest, ValidatePromoCodeRequest, ValidatePromoCodeResponse } from '../types';

class PromoCodeService {
  // ── Public / Checkout endpoints ─────────────────────────────────────
  async validatePromoCode(request: ValidatePromoCodeRequest): Promise<ValidatePromoCodeResponse> {
    const response = await api.post<ValidatePromoCodeResponse>('/api/public/promocodes/validate', request);
    return response.data;
  }

  // ── Admin endpoints ─────────────────────────────────────────────────
  async getAllPromoCodes(): Promise<PromoCode[]> {
    const response = await api.get<PromoCode[]>('/api/admin/promocodes');
    return response.data;
  }

  async createAdminPromoCode(request: PromoCodeRequest): Promise<PromoCode> {
    const response = await api.post<PromoCode>('/api/admin/promocodes', request);
    return response.data;
  }

  async updateAdminPromoCode(id: string, request: PromoCodeRequest): Promise<PromoCode> {
    const response = await api.put<PromoCode>(`/api/admin/promocodes/${id}`, request);
    return response.data;
  }

  async toggleAdminPromoCodeStatus(id: string): Promise<PromoCode> {
    const response = await api.patch<PromoCode>(`/api/admin/promocodes/${id}/toggle`);
    return response.data;
  }

  async deleteAdminPromoCode(id: string): Promise<void> {
    await api.delete(`/api/admin/promocodes/${id}`);
  }

  // ── Organizer endpoints ─────────────────────────────────────────────
  async getOrganizerPromoCodes(): Promise<PromoCode[]> {
    const response = await api.get<PromoCode[]>('/api/organizer/promocodes');
    return response.data;
  }

  async createOrganizerPromoCode(request: PromoCodeRequest): Promise<PromoCode> {
    const response = await api.post<PromoCode>('/api/organizer/promocodes', request);
    return response.data;
  }

  async updateOrganizerPromoCode(id: string, request: PromoCodeRequest): Promise<PromoCode> {
    const response = await api.put<PromoCode>(`/api/organizer/promocodes/${id}`, request);
    return response.data;
  }

  async toggleOrganizerPromoCodeStatus(id: string): Promise<PromoCode> {
    const response = await api.patch<PromoCode>(`/api/organizer/promocodes/${id}/toggle`);
    return response.data;
  }

  async deleteOrganizerPromoCode(id: string): Promise<void> {
    await api.delete(`/api/organizer/promocodes/${id}`);
  }
}

const promoCodeService = new PromoCodeService();
export default promoCodeService;

File: ticket-booking-system/src/main/java/com/ticket/ticket_booking_system/service/PromoCodeService.java
package com.ticket.ticket_booking_system.service;

import java.util.List;
import java.util.UUID;

import com.ticket.ticket_booking_system.dto.request.PromoCodeRequest;
import com.ticket.ticket_booking_system.dto.request.ValidatePromoCodeRequest;
import com.ticket.ticket_booking_system.dto.response.PromoCodeResponse;
import com.ticket.ticket_booking_system.dto.response.ValidatePromoCodeResponse;

public interface PromoCodeService {

    // Admin methods
    List<PromoCodeResponse> getAllPromoCodes();

    PromoCodeResponse createAdminPromoCode(PromoCodeRequest request, String adminEmail);

    PromoCodeResponse updateAdminPromoCode(UUID id, PromoCodeRequest request, String adminEmail);

    void deleteAdminPromoCode(UUID id, String adminEmail);

    PromoCodeResponse togglePromoCodeStatus(UUID id, String userEmail);

    // Organizer methods
    List<PromoCodeResponse> getOrganizerPromoCodes(UUID organizerId);

    PromoCodeResponse createOrganizerPromoCode(PromoCodeRequest request, UUID organizerId, String organizerEmail);

    PromoCodeResponse updateOrganizerPromoCode(UUID id, PromoCodeRequest request, UUID organizerId, String organizerEmail);

    void deleteOrganizerPromoCode(UUID id, UUID organizerId, String organizerEmail);

    PromoCodeResponse toggleOrganizerPromoCodeStatus(UUID id, UUID organizerId, String organizerEmail);

    // Public / Validation methods
    ValidatePromoCodeResponse validateAndCalculateDiscount(ValidatePromoCodeRequest request);

    void incrementUsage(String code);
}
